🏃 Loosen webhook validation on control plane to allow etcd changes

chuckha opened the pull request:
Signed-off-by: Chuck Ha <chuckh@vmware.com>

**What this PR does / why we need it**:
This PR loosens the web hook restriction allowing for changes to the etcd field but disallowing any other ClusterConfiguation field change. 

Allows setting the etcd local image metadata after creation
Allows unsetting the etcd local image metadata after having set it
Denies changing any non-etcd local image metadata (either setting or unsetting)

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Related to #2543

vincepri commented on controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go:
```diff
@@ -65,17 +72,81 @@ func (in *KubeadmControlPlane) ValidateUpdate(old runtime.Object) error {
+	}
+	if !reflect.DeepEqual(in.Spec.KubeadmConfigSpec.ClusterConfiguration.ClusterName, prev.Spec.KubeadmConfigSpec.ClusterConfiguration.ClusterName) {
+		allErrs = append(allErrs, forbiddenSpecError(kubeadmConfigSpec, clusterConfiguration, "clusterName"))
+	}
```
This seems we might need to keep it up to date for the future? I wonder if we could take an approach saying like we don't allow any changes except for X? We could potentially use unstructured to compare changes between json objects rather than do field-by-field

chuckha replied:
It will have to change when the types go from kubeadm v1beta1 to v1beta2 (which they should soon since we're getting deprecation warnings now 😄).

Yeah, the redundancy is super annoying. To expand on your suggestion, I'm assuming unstructured might have some kind of diff functionality we could use here and then we'd diff the two objects, see which fields changed and then deny or allow based on that?

I can see if unstructured would provide us with that information

vincepri replied:
I suggested unstructured because we can easily remove "paths" from the objects that we'll allow mutations regardless, but that's not strictly necessary.

We might be able to leverage https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/client/patch.go#L87 to generate a patch and see if there are any diffs.

chuckha replied:
reasonable. Appreciate the pushback against this wall of copy paste, i'll find a better solution! 😅

vincepri replied:
btw I'm ok doing this for now, opening an issue and improve in v0.3.x

vincepri commented:
/milestone v0.3.0

detiber commented on controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go:
```diff
@@ -60,22 +61,49 @@ func (in *KubeadmControlPlane) ValidateCreate() error {
+		{spec, kubeadmConfigSpec, clusterConfiguration, "etcd", "local", "imageTag"},
+		{spec, "infrastructureTemplate", "name"},
+		{spec, "replicas"},
+	}
```
I'm not entirely sure how the status subresource is handled wrt admission webhooks, but do we need to worry about this blocking status updates?

vincepri replied:
Similar to above, maybe we should focus only on spec for now, wydt? I can think that for example in we might want to disallow updates to the cluster-name label in metadata in the future, but we can probably handle that separately

chuckha replied:
I'm not sure either, I'll dig into that and figure it out

chuckha replied:
This is what indicates what will or won't get sent through validation: https://github.com/kubernetes-sigs/cluster-api/blob/master/controlplane/kubeadm/config/webhook/manifests.yaml#L45-L54

This will only send the spec and none of the sub resources including status so status does not go through validation.

This holds true with the tests that I've run that update the status of the KCP even though status is not a valid field according to the web hook validator.

chuckha replied:
probably also helpful to post the documentation links: https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#webhook-configuration

detiber replied:
Awesome, thanks for clarifying.

chuckha commented:
/test pull-cluster-api-test

randomvariable commented:
/test pull-cluster-api-capd-e2e

chuckha commented:
/assign @vincepri 

This is ready to go if you're good with it

ncdc commented on controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go:
```diff
@@ -84,6 +109,56 @@ func (in *KubeadmControlPlane) ValidateUpdate(old runtime.Object) error {
+}
+
+func forbiddenSpecError(path ...string) *field.Error {
+	return field.Forbidden(field.NewPath(path[0], path[1:]...), "cannot be modified")
```
Probably safest to check the length to avoid index out of bounds errors?

ncdc commented on controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook_test.go:
```diff
@@ -278,6 +358,120 @@ func TestKubeadmControlPlaneValidateUpdate(t *testing.T) {
+			kcp:       scheduler,
+		},
+		{
+			name:      "should fail when making a change to the cluster config's dns",
```
We'll need this to be allowed as part of #2574, either here or as part of @wfernandes's PR, just FYI

ncdc commented on controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go:
```diff
@@ -60,22 +61,46 @@ func (in *KubeadmControlPlane) ValidateCreate() error {
-			),
-		)
-	}
+	patch := client.MergeFrom(prev)
```
I _think_ this is pulling in spdystream and goproxy, maybe. This code is pretty straight-forward if you wanted to try to eliminate those dependencies - here's what `client.MergeFrom` and `patch.Data` do:

```go
originalJSON, err := json.Marshal(s.from)
	if err != nil {
		return nil, err
	}

	modifiedJSON, err := json.Marshal(obj)
	if err != nil {
		return nil, err
	}

	return jsonpatch.CreateMergePatch(originalJSON, modifiedJSON)
```

chuckha replied:
for sure, always happy to remove dependencies

vincepri commented:
LGTM pending @ncdc's comments, looks great thanks for doing this @chuckha !

chuckha commented:
@vincepri addressed the feedback. I'm not completely sure on the types of errors I chose to return (apierrors.InternalError) but they seemed to fit the best

vincepri commented:
InternalError does seems the most appropriate, I'd expect users to file a bug if any of that happens
